### PR TITLE
Module code add validation

### LINKED ISSUE #93: Add validation method to ModuleCode class
The current `ModuleCode` class does not have validation for constructor which makes it error-prone. Since `ModuleRoleMap` class depends on this method as well, the implementation of the method is compulsory.

### PR DESCRIPTION
**Pull request by MAOXIONGKAI**
Closes #93 
Implement validation check for `ModuleCode` class such that:
1. The regex pattern used is `[A-Za-z]+[0-9]+[A-Za-z]?` which matches at least one letter at the start of the name, followed by at least one number, and lastly end with optional letters to fit into the module code that y1 cs student may need to input.
2. Test cases have been included to make sure that `ModuleCode` class can only be instantiated when input string matches the above mentioned pattern.
3. `.trim()` method is added in the constructor to trim off the leading and trailing spaces in the input so that the regex pattern and validation method does not have to check for this kind of edge cases.
4. `.toUpperCase()` method is added in the constructor to convert valid inputs to upper case, to ensure consistent styling and easy search of module code within the program.

### CODE REVIEW

File: src/main/java/seedu/address/model/person/ModuleCode.java [hunk 1]
```diff
@@ -1,12 +1,25 @@
...
+            + "The accepted pattern is [Alphabets][Numbers][Alphabets], where the ending [Alphabets] are optional.\n"
+                    + "Note that input check is case-insensitive and program will automatically convert valid inputs\n"
+                    + "to upper case letters for consistent format and easy search of module codes."
+                    + "For example: MA1522, cs2106, CS1101S, is1108, CS2103t are all considered valid inputs.";
+
+    /*
+     * The input should consist of only alphanumeric characters without spaces, should not contain
+     * space, and should not be blank.
+     */
+    public static final String VALIDATION_REGEX = "[A-Za-z]+[0-9]+[A-Za-z]?";
```
Comment (lithops-zty) [inline src/main/java/seedu/address/model/person/ModuleCode.java:22]: I realize the suffix length may be more than 1 for a lot of courses (472 courses in this AY). Among them, there are courses that may be taken by Y1 CS students, for example 'CFG2002ID', 'CFG2002KH', 'CFG2002PH', 'CFG2002TH' and 'CFG2002VN' of the Global Industry Insights series where the suffix represents the destination country. Maybe we can change `?` to `{0, 2}` to allow such module codes.
```suggestion
    public static final String VALIDATION_REGEX = "[A-Za-z]+[0-9]+[A-Za-z]{0, 2}";
```

Comment (MAOXIONGKAI) [reply src/main/java/seedu/address/model/person/ModuleCode.java]: Hi, thanks for the review! I confused myself with the meaning of `?` in regex and didn't write test case to cover this kind of edge cases. Thank you so much for spotting that! However, instead of `{0, 2}`, would it be better to replace it with `*`(match 0 or more) instead? Because in some rare cases there may still be suffix with more letters and I think it make sense to handle that kinds of inputs as well. Thank you!
